orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,208 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
+  for (auto i = from_stage; i < to_stage; ++i) {
+    // processing send node in cut i
+    auto& send_attributes = send_nodes[i]->GetMutableAttributes();
+    auto& send_element_types = send_attributes["element_types"];
+    send_element_types.add_ints(static_cast<int64_t>(dtype));
+    send_nodes[i]->MutableInputDefs().push_back(current_node_arg);
+    send_nodes[i]->MutableInputArgsCount().back()++;
+
+    // Create a new node_arg for the recv, as the new node_arg from recv node should possess a differnet id
+    // than the one in send
+    auto& new_node_arg = CreateNodeArg(graph, current_node_arg);
+    new_node_args.push_back(&new_node_arg);
+    current_node_arg = &new_node_arg;
+
+    // process recv node in cut i
+    auto& recv_attributes = recv_nodes[i]->GetMutableAttributes();
+    auto& recv_element_types = recv_attributes["element_types"];
+    recv_element_types.add_ints(static_cast<int64_t>(dtype));
+    recv_nodes[i]->MutableOutputDefs().push_back(current_node_arg);
+  }
+
+  // update the consumer node's input if the node's group is not in the first partition
+  for (size_t i = 1u; i < node_groups.size(); ++i) {
+    ORT_ENFORCE(node_groups[i].stage_id > from_stage, "node group id (", node_groups[i].stage_id,
+                ") is less than first stage id (", from_stage, "). ");
+    size_t new_node_arg_index = node_groups[i].stage_id - from_stage - 1;
+    for(auto node : node_groups[i].nodes){
+      for (auto& input_node : node->MutableInputDefs()) {
+        if (input_node == initializer) {
+          input_node = new_node_args[new_node_arg_index];
+          break;
+        }
+      }
+    }
+  }
+  return Status::OK();
+}
+
+void TraverseGraphWithConnectedElement(Graph& graph,
+                                       const Node* startNode,
+                                       std::set<const Node*>& visited_nodes,
+                                       std::set<const NodeArg*>& visited_inputs,
+                                       std::set<const NodeArg*>& visited_outputs) {
+  visited_nodes.clear();
+  visited_inputs.clear();
+  visited_outputs.clear();
+
+  std::queue<const Node*> node_queue;
+  node_queue.push(startNode);
+
+  while (!node_queue.empty()) {
+    auto node = node_queue.front();
+    node_queue.pop();
+    if (visited_nodes.count(node) == 0) {
+      visited_nodes.insert(node);
+      std::vector<const Node*> connected_nodes;
+      ORT_THROW_IF_ERROR(FindAllConnectedNodes(graph, node, connected_nodes, visited_inputs, visited_outputs));
+
+      for (auto n : connected_nodes) {
+        ORT_ENFORCE(n != nullptr, "Found nullptr in searching for connected nodes");
+        node_queue.push(n);
+      }
+    }
+  }
+}
+
+// If an initializer is shared across partitions, instead of creating a separate all_reduce op to
+// sync with those tensors in selected partitions, we save only one copy of that initializer in
+// the very first partition it appears, and pass that data down to all following partitions
```

Comment:
High-level question: Does the current rewrite pass the shared initializer along with the activations on every microbatch, or does it only do it after each weight update?

Reply:
The shared initializer is passed along with activations on every microbatch. So the last partition don't store the tensor and thus ease the memory pressure from the last partition. 

---
In reply to: [444999349](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r444999349) [](ancestors = 444999349)